src/librustc_data_structures/graph_algorithms/loop_tree/tree.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use super::super::Graph;
+use super::super::node_vec::NodeVec;
+
+pub struct LoopTree<G: Graph> {
+    loop_ids: NodeVec<G, Option<LoopId>>,
+    loop_infos: Vec<LoopInfo<G>>,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct LoopId {
+    index: usize
+}
+
+struct LoopInfo<G: Graph> {
```

Comment:
I'm not sure if we need the loop-tree for anything yet? (though I think we may want it eventually...) Might be better to hold off, particularly since this data structure is the least standard of the set and I think I was kind of experimenting with how it ought to be setup.